### ISSUE TITLE
Refactor the Scalastyle Plugin

### ISSUE THREAD

**Issue by BalmungSan** [mwz/sonar-scala#35]
The idea is to do something similar as #34, in order to remove all SonarQube deprecations and thus finally closing #7.

The first thing to do, is to carefully review the code, in order to decide if we should adapt the actual code to the new sonar API, or if its cleaner and cheaper to rewrite the plugin almost from scratch.

This issue will serve as discussion space, for planning the refactor before doing it.

**Comment by BalmungSan** [comment]
Well, the time to tackle this has come.

IMHO, I see no value in only refactoring the rules repository, and I think we should refactor the entire module at once. Of course doing a big one PR to change all it's not a good idea _(we learned that from the **Scoverage** refactor)_, so we probably will do something similar of what we did for **Scapegoat**, tackling this part by part, without removing the old module until the new one is ready.

I haven't given a detailed review to this, but this is a high level task list:

+ SBT code generation task, to create a file with all the **Scalastyle** rules.
+ **Scalastyle** rules repository.
+ **Scalastyle** quality profile.
+ **Scalastyle** sensor _(Includes feature flag)_.
+ Docs, examples & release.

What do you think @mwz ?

**Comment by mwz** [comment]
It looks like a good list to me!

**Comment by mwz** [comment]
> I think we should do something similar of what we did for scapegoat, using fast-classpath-scanner to get all classes that inherit from Checker (Or maybe from Scalariform Checker).

I don't think that we'll need to use the `fast-classpath-scanner` for Scalastyle, because everything we need is actually located in the xml config files in the resources [here](https://github.com/scalastyle/scalastyle/tree/master/src/main/resources).

**Comment by BalmungSan** [comment]
> "I don't think that we'll need to use the fast-classpath-scanner for Scalastyle, because everything we need is actually located in the xml config files in the resources here."

Oh that's cool 🎉
